### PR TITLE
dist/tools: test automation framework

### PR DESCRIPTION
**Pull request by cgundogan**
This PR introduces a very simple test automation framework that @authmillenon and me worked on for a couple of hours.

The idea is to have a very flexible test framework that runs the tests in `RIOTBASE/tests/` and is extensible to cover not only compile tests but also more sophisticated network tests (later)

### CODE REVIEW

File: dist/tools/autotest/autotest/default.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# directory for more details.
+
+from __future__ import print_function
+import pexpect
+from subprocess import check_output, STDOUT, CalledProcessError
+import os
+import shlex
+
+RIOT_BASE = '../../../'
+TESTS_BASE = RIOT_BASE + 'tests/'
```
Comment (authmillenon) [inline dist/tools/autotest/autotest/default.py]: I would let the user let set at least `RIOT_BASE`

File: dist/tools/autotest/autotest/__init__.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        if self.port:
+            env['PORT'] = self.port
+        if self.serial:
+            env['SERIAL'] = self.serial
+        return env
+
+class TestCase(object):
+    def __init__(self, name, build_strategy = None, test_strategy = None):
+        self.name = name
+        if build_strategy:
```
Comment (kaspar030) [inline dist/tools/autotest/autotest/__init__.py]: self.build_action = build_strategy or None

File: dist/tools/autotest/autotest/__init__.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            self.build_action = build_strategy
+        else:
+            self.build_action = None
+
+        if test_strategy:
```
Comment (kaspar030) [inline dist/tools/autotest/autotest/__init__.py]: same

File: dist/tools/autotest/autotest/__init__.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            self.test_action = test_strategy
+        else:
+            self.test_action = None
+
+    def build(self):
+        if self.build_action:
+            print(' - {} '.format(colorize('Building', TermColor.purple)), end='')
+            if self.build_action.build(self.name):
+                print('{}'.format(colorize('successfull', TermColor.green)))
```
Comment (kaspar030) [inline dist/tools/autotest/autotest/__init__.py]: successful

File: dist/tools/autotest/autotest/__init__.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                return True
+            else:
+                print('{}'.format(colorize('failed', TermColor.red)))
+                return False
+
+    def test(self):
+        if self.test_action:
+            print(' - {} '.format(colorize('Testing', TermColor.purple)), end='')
+            if self.test_action.test(self.name):
+                print('{}'.format(colorize('successfull', TermColor.green)))
```
Comment (kaspar030) [inline dist/tools/autotest/autotest/__init__.py]: successful

File: .gitignore [hunk 6]
```diff
@@ -28,6 +28,9 @@ cachegrind.out*
 tags
 # GDB initialization scripts
 .gdbinit
+# Python files
+*.pyo
```
Comment (Kijewski) [inline .gitignore]: \+ *.pyc
\+ *.pyd

Comment (authmillenon) [inline .gitignore]: don't know what I meant by *.pyo though. I was in a hurry ^^.

File: dist/tools/autotest/run.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        ('coap', BOARDS, DefaultBuildStrategy, None),
+        ('fmt_print', BOARDS, DefaultBuildStrategy, DefaultTestStrategy),
+    ]
+
+def run_tests(tests):
+    for test in tests:
+        tc = autotest.TestCase(test[0],
+                               build_strategy=test[2](test[1]) if test[2] else None,
+                               test_strategy=test[3](test[1]) if test[3] else None)
+        tc.execute()
```
Comment (authmillenon) [inline dist/tools/autotest/run.py]:
```diff
         ('fmt_print', BOARDS, DefaultBuildStrategy, DefaultTestStrategy),
     ]
 
+def run_test(name, boards, build_strategy = None, test_strategy = None):
+        tc = autotest.TestCase(name, build_strategy(boards) if build_strategy else None,
+                               test_strategy=test_strategy(boards) if test_strategy else None)
+        tc.execute()
+
 def run_tests(tests):
     for test in tests:
-        tc = autotest.TestCase(test[0],
-                               build_strategy=test[2](test[1]) if test[2] else None,
-                               test_strategy=test[3](test[1]) if test[3] else None)
-        tc.execute()
+        run_test(*test)
 
 if __name__ == '__main__':
     run_tests(TEST_CASES)
```

Might be nicer to look at for some people.

File: .gitignore (hunk 6 above)
Comment (cgundogan) [inline .gitignore]: > When the Python interpreter is invoked with the -O flag, optimized code is generated and stored in ‘.pyo’ files.

**Comment by cgundogan** [comment]
addressed comments

**Comment by authmillenon** [comment]
Some additional thoughts I had with a little distance:
* Execution of just a subset of tests would be nice
* Execution for just a subset would also be nice
* I just realized that additional parameters to the strategies' constructors are not supported by this version of the framework (your previous version did however)
* grouping of boards for defining which board should communicate with each other for network functionality tests would also be nice (though that might be able to be constructed in the specific TestStrategy).

I think we can make the first point easier to we conceptualize this framework more as a library, so that a version of the `run.py` script would be residing in the test's directory itself. This way we would also circumvent another `Makefile.dep` situation in this file.

The second (and maybe the fourth point, too) could be realized by letting the boards be created by a list of arguments given to the script (at least `port` and `serial` will be different for most users and boards anyway). I'm thinking qemu-style parameters here, e.g.: 

```bash
./run.py --board native \
         -b native,port=tap1 \
         --board samr21-xpro,port=/dev/ttyACM0,serial=ATML2127031800001935 \
         --board samr21-xpro,port=/dev/ttyACM1,serial=ATML2127031800002743`
```

**Comment by cgundogan** [comment]
> ```
./run.py --board native \
         -b native,port=tap1 \
         --board samr21-xpro,port=/dev/ttyACM0,serial=ATML2127031800001935 \
         --board samr21-xpro,port=/dev/ttyACM1,serial=ATML2127031800002743`
```

I am missing the mapping of the boards to the specific testcases here. While one test might need several boards, other do need just one

**Comment by cgundogan** [comment]
>
Execution of just a subset of tests would be nice
Execution for just a subset would also be nice

I have to admit, I do not see the difference between those two. Currently, it is possible to run a `subset` of tests by commenting out the tests from the `TEST_CASES` array

**Comment by cgundogan** [comment]
It would also be possible to defined more `TEST_CASES` arrays and use the one which suits the purpose

**Comment by authmillenon** [comment]
> I am missing the mapping of the boards to the specific testcases here. 

Granted it's no mapping to the TestCases, but to the applications (given, that `run.py` is per application), but I think that would suffice. If a certain test case needs another test configuration it can pick from the selection of given boards.

> While one test might need several boards, others do need just one.

But the number of tests that only run on only one board is comparably small.

> I have to admit, I do not see the difference between those two. Currently, it is possible to run a subset of tests by commenting out the tests from the TEST_CASES array
...
> It would also be possible to defined more TEST_CASES arrays and use the one which suits the purpose

I thought we are trying to automate things here. Having several scripts to run is far easier on a CI server to do, than commenting out stuff ;-). Since Python is pretty lenient in the difference between a module and a script there is still the possibility to collect all the TEST_CASES from all the test scripts and execute them somewhere central if need for this arises.

**Comment by OlegHahm** [comment]
What's the latest? @A-Paul, did you have a look?

**Comment by cgundogan** [comment]
I think @authmillenon was working on another test framework that supersedes this one. Am I right, @authmillenon ?

**Comment by LudwigKnuepfer** [comment]
No documentation?

**Comment by miri64** [comment]
Let's rethink that more thoroughly. The current solution works to an extend and I have somewhat of a solution in the pipe. Will present it as soon as I have something substantial.

**Pull request closed without merging**